### ISSUE TITLE
D4PG's vmin and vmax paramaters

### ISSUE THREAD

**Issue by rasoolfa** [deepmind/acme#57]
Hi @fastturtle

Would it be possible to share more info about hyper-parameters D4PG? For example, D4PG paper doesn't include information about all of important hyperparameters in the paper such as **vmin** and **vmax** . Although there is a [note](https://github.com/deepmind/acme/tree/799f76fffe8dde4e6bbf3c6cb1c0260cf7ff00aa/acme/agents/tf/d4pg) says how to find them, it can still lead to very different results and it is kinda unclear. 
If it is not possible to share them, can you clarify a bit the note about vmin and vmax? 
"A good rule of thumb is to set vmax to the discounted sum of the maximum instantaneous rewards for the maximum episode length;" What policy should be used to gather this info and when? random policy? or we should train an agent for a while with another method and then use that policy it to get maximum reward? how this vmax/vmin related to the final an agent's returns and optimal policy? For example, final return of Acrobot(Swingup) is ~400 (Figure 2 in the paper), so should vmax be very close to 400?

Thanks.

**Comment by bshahr** [comment]
Hi rasoolfa,

Thanks for your question! Indeed the problem of setting the vmin/vmax in D4PG (and C51) is not obvious especially when the environment's reward signal is completely unknown. Thankfully, in the `dm_control` suite the rewards are all standardized so let me use that as an example for clarification.

The instantaneous rewards (reward obtained at each timestep) in any `dm_control` task is bound in `(-1,1)`. If the distributional Q function is meant to fit the distribution of discounted returns, ie the discounted sum of instantaneous rewards, the maximum possible value is a discounted sum of 1s. Therefore the maximum possible return is simply the value of the geometric series with ratio `gamma = 0.99` as set in all our experiments, which leads to a `vmax` of 100. Historically, a value of 150 was used to give it a bit of slack but you could probably get away without it. (Let us know if you try it!)

As for your specific question about `acrobot:swingup`: first of all, you're missing the discounting in your calculation of `vmax`; secondly, just because we haven't found a policy that reaches higher returns doesn't mean that there doesn't exist one, therefore our Q function had better be able to evaluate such a policy. This is why we use the theoretical limits set by the `dm_control` designers and fix `vmin|vmax` for all tasks therein.

Hope that helped!

Bobak

**Comment by rasoolfa** [comment]
Thanks for your detailed response, it is very helpful indeed.
